### PR TITLE
Add buffered writer

### PR DESCRIPTION
**Pull request by yinan1048576**
The buffered writer adopts a signature identical to `write_cb`,
so that it can be plugged into anywhere `write_cb` appears.
Currently it's only used in `malloc_stats_print` when both the
`write_cb` parameter and `je_malloc_message` are not set - without
buffering (as is the case before this change) it would invoke
`malloc_write_fd` for every single line in the stats output.

### CODE REVIEW

**Comment by yinan1048576** [comment]
It is also feasible to enable buffered writing by default for `malloc_stats_print`. I imagine there might be cases where the call back is a "`grep`"-like function, in which case buffering might not be what the caller wants, but most of the times the call back should just be some writing function and buffering can be very helpful.

**Comment by yinan1048576** [comment]
Separated out the change to `malloc_stats_print` as a subsequent change so as to improve modularity. Also improved & added more comments.

**Comment by yinan1048576** [comment]
Rebase, and put unit tests into the same PR.

File: src/malloc_io.c [hunk 1]
```diff
@@ -664,6 +664,33 @@ malloc_printf(const char *format, ...) {
 	va_end(ap);
 }
 
+void
+buf_writer_flush(buf_writer_arg_t *arg) {
+	arg->buf[arg->buf_end] = '\0';
```
Comment (interwq) [inline src/malloc_io.c]: nit: assert end < size.

File: src/malloc_io.c [hunk 2]
```diff
@@ -664,6 +664,33 @@ malloc_printf(const char *format, ...) {
...
+
+void
+buffered_write_cb(void *buf_writer_arg, const char *s) {
+	buf_writer_arg_t *arg = (buf_writer_arg_t *)buf_writer_arg;
+	size_t i, n, slen;
+	for (i = 0, slen = strlen(s); i < slen; i += n) {
+		if (arg->buf_end == arg->buf_size) {
+			buf_writer_flush(arg);
+		}
+		n = slen - i < arg->buf_size - arg->buf_end ?
```
Comment (interwq) [inline src/malloc_io.c]: nit: add a `buf_remain = arg->buf_size - arg->buf_end` to help readability.

File: src/malloc_io.c [hunk 3]
```diff
@@ -664,6 +664,33 @@ malloc_printf(const char *format, ...) {
...
+	arg->write_cb(arg->cbopaque, arg->buf);
+	arg->buf_end = 0;
+}
+
+void
+buffered_write_cb(void *buf_writer_arg, const char *s) {
+	buf_writer_arg_t *arg = (buf_writer_arg_t *)buf_writer_arg;
+	size_t i, n, slen;
+	for (i = 0, slen = strlen(s); i < slen; i += n) {
+		if (arg->buf_end == arg->buf_size) {
```
Comment (interwq) [inline src/malloc_io.c]: Move this to the end of the loop?

File: test/unit/buf_writer.c [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+TEST_BEGIN(test_buf_write) {
+	char s[UNIT_MAX + 1];
+	size_t unit, n_unit, remain, i;
+	size_t arg = 1;
+	buf_writer_arg_t test_buf_arg =
+	    {test_write_cb, &arg, test_buf, TEST_BUF_SIZE - 1, 0};
+	memset(s, 'a', UNIT_MAX);
+	for (unit = UNIT_MAX; /* include 0 for edge case */; --unit) {
```
Comment (interwq) [inline test/unit/buf_writer.c]: nit: make `unit` signed and check `unit >= 0`?

File: test/unit/buf_writer.c [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+		/* unit keeps decreasing, so strlen(s) is always unit. */
+		s[unit] = '\0';
+		for (n_unit = 1; n_unit <= 3; ++n_unit) {
+			test_write_len = 0;
+			remain = 0;
+			for (i = 1; i <= n_unit; ++i) {
+				/* A toy pseudo-random number generator */
```
Comment (interwq) [inline test/unit/buf_writer.c]: Any particular reason to not use the prng functions?

File: src/malloc_io.c (hunk 3 above)
Comment (yinan1048576) [reply src/malloc_io.c]: The logic might be slightly different. Having the test and the flush upfront makes flushing lazy: if the input exactly fits the entire buffer, the flush won't take place until the next call of `buffered_write_cb()` or `buf_writer_flush()`. (My unit tests are particularly testing for this behavior.) The advantage is that the final `buf_writer_flush()` call would never unnecessarily call the real `write_cb()` one more time with an empty buffer. (Of course I can guard against an empty buffer in the `buf_writer_flush()` function, but that feels a bit redundant.)

**Comment by yinan1048576** [comment]
Revised according to feedback. Added more asserts and better variable namings.

File: src/malloc_io.c (hunk 3 above)
Comment (interwq) [reply src/malloc_io.c]: Ah I see. Didn't think of the final `buf_writer_flush()` call. Makes sense.
